### PR TITLE
Allow API socket deactivation by passing no-api parameter

### PR DESCRIPTION
**Pull request by lauralt**
Signed-off-by: Laura Loghin <lauralg@amazon.com>

## Reason for This PR

We want to allow the user to not use the API socket at all, if he configures the microvm with a configuration file and doesn't need to make post-boot operations (PATCH, GET calls).

Fixes: #1165

## Description of Changes

Added new command line parameter for Firecracker, named `--no-api` which will disable API server thread. Also, to avoid ambiguity, now the main thread is the VMM thread and the API thread is the one that is spawned from `main()`.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria. Where there are two options, keep one.]`  
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is
      clearly provided. 
- [x] The description of changes is clear and encompassing.
- [x] The required
      doc changes are included in this PR. Docs in scope are all `*.md` files
      located either in the repository root, or in the `docs/` directory.
- [x] Code-level documentation for touched
      code is included in this PR.
- [x] No API changes are included in this PR.
- [x] The changes in
      this PR have user impact and have been added to the `CHANGELOG.md` file.

### CODE REVIEW

File: tests/integration_tests/functional/test_cmd_line_start.py [hunk 1]
```diff
@@ -46,7 +34,57 @@ def test_config_json_start(test_microvm_with_ssh, vm_config_file):
...
     test_microvm.create_jailed_resource(vm_config_path, create_jail=True)
     test_microvm.config_file = os.path.basename(vm_config_file)
+
+
+@pytest.mark.parametrize(
+    "vm_config_file",
+    ["framework/vm_config.json"]
+)
+def test_config_start_with_api(test_microvm_with_ssh, vm_config_file):
+    """Test that microvm that was configured from file has booted."""
```
Comment (sandreim) [inline tests/integration_tests/functional/test_cmd_line_start.py]:
```suggestion
    """Test if a microvm configured from file boots successfully."""
```

File: src/main.rs [hunk 2]
```diff
@@ -155,40 +166,74 @@ fn main() {
...
+
+        let event_fd = EventFd::new()
+            .map_err(Error::Eventfd)
+            .expect("'cannot create Eventfd.");
+
+        let api_event_fd = event_fd.try_clone().expect("cannot clone Eventfd.");
+        let api_shared_info = shared_info.clone();
+
+        thread::Builder::new()
+            .name("fc_api".to_string())
```
Comment (sandreim) [inline src/main.rs]: nit: I prefer to use to_owned() instead of to_string() in this case.

File: src/main.rs [hunk 3]
```diff
@@ -155,40 +166,74 @@ fn main() {
...
+            .spawn(move || {
+                let server = ApiServer::new(mmds_info, shared_info, to_vmm, event_fd)
```
Comment (sandreim) [inline src/main.rs]: Why not:

```suggestion
match ApiServer::new(mmds_info, shared_info.clone(), to_vmm, event_fd)
                .expect("Cannot create API server")
                .bind_and_run(
```

File: src/main.rs (hunk 2 above)
Comment (lauralt) [reply src/main.rs]: Done

File: src/main.rs (hunk 3 above)
Comment (lauralt) [reply src/main.rs]: Done

File: tests/integration_tests/functional/test_cmd_line_start.py (hunk 1 above)
Comment (lauralt) [reply tests/integration_tests/functional/test_cmd_line_start.py]: Done

File: vmm/src/lib.rs [hunk 4]
```diff
@@ -2176,83 +2176,76 @@ impl PartialEq for VmmAction {
...
+    let mut vmm = Vmm::new(
+        api_shared_info,
+        &api_event_fd,
+        from_api,
+        seccomp_level,
+        Kvm::new().expect("Error creating the Kvm object"),
+    )
+    .expect("Cannot create VMM");
+
+    if let Some(json) = config_json {
```
[the remark this answers is not in the document]
Comment (lauralt) [reply vmm/src/lib.rs]: Looks nice, modified accordingly.

File: src/main.rs [hunk 5]
```diff
@@ -155,40 +166,69 @@ fn main() {
...
         id: instance_id,
         vmm_version: crate_version!().to_string(),
     }));
-    let mmds_info = MMDS.clone();
-    let (to_vmm, from_api) = channel();
-    let server =
-        ApiServer::new(mmds_info, shared_info.clone(), to_vmm).expect("Cannot create API server");
-
-    let api_event_fd = server
-        .get_event_fd_clone()
```
[the remark this answers is not in the document]
Comment (lauralt) [reply src/main.rs]: Done

File: tests/integration_tests/build/test_coverage.py [hunk 6]
```diff
@@ -19,7 +19,7 @@
 
 import host_tools.cargo_build as host  # pylint: disable=import-error
 
-COVERAGE_TARGET_PCT = 85.9
+COVERAGE_TARGET_PCT = 85.8
```
[the remark this answers is not in the document]
Comment (lauralt) [reply tests/integration_tests/build/test_coverage.py]: I've added in a new commit some tests for vsock to increase coverage, as I find it pretty difficult to test the newly added lines in `main()`.

File: src/main.rs [hunk 7]
```diff
@@ -155,40 +166,69 @@ fn main() {
...
+        vmm::start_vmm(
+            shared_info,
+            event_fd,
+            from_api,
+            seccomp_level,
+            vmm_config_json,
+        );
+    } else {
+        let (to_vmm, from_api) = channel();
+        let mmds_info = MMDS.clone();
```
[the remark this answers is not in the document]
Comment (lauralt) [reply src/main.rs]: Done, I'll also open an issue.

File: api_server/src/lib.rs [hunk 8]
```diff
@@ -79,12 +79,13 @@ impl ApiServer {
         mmds_info: Arc<Mutex<Mmds>>,
         vmm_shared_info: Arc<RwLock<InstanceInfo>>,
         api_request_sender: mpsc::Sender<Box<VmmAction>>,
+        efd: EventFd,
```
[the remark this answers is not in the document]
Comment (lauralt) [reply api_server/src/lib.rs]: Done

File: src/main.rs [hunk 9]
```diff
@@ -155,40 +166,69 @@ fn main() {
...
-            _ => panic!(
-                "Failed to communicate with the API socket: {:?}",
-                Error::Io(inner)
-            ),
-        },
-        Err(eventfd_err @ Error::Eventfd(_)) => {
-            panic!("Failed to open the API socket: {:?}", eventfd_err)
-        }
+
+    if no_api {
```
[the remark this answers is not in the document]
Comment (lauralt) [reply src/main.rs]: Done

**Pull request merged**